sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -369,6 +409,68 @@ def _get_environment(
         env = self._environment_from_row(row)
         return env
 
+    @transactional()
+    def migrate(self) -> None:
+        super().migrate()
+
+    @transactional()
+    def _migrate_rows(self) -> None:
+        all_snapshots = self._get_snapshots(lock_for_update=True)
+        environments = self.get_environments()
+        snapshots_by_version = defaultdict(list)
+
+        for snapshot in all_snapshots.values():
+            snapshots_by_version[(snapshot.name, snapshot.version)].append(snapshot)
+
+        new_snapshots = set()
+
+        for environment in environments:
```

Comment:
How about snapshots that are not part of any environments but which haven't been garbage collected yet?